File: Facade_Pattern/fp.cpp
#include <algorithm>
#include <iostream>
#include <memory>
#include <vector>

class DvdPlayer {
   public:
    void on() { std::cout << "Dvd player is on\n"; }
    void off() { std::cout << "Dvd player is off\n"; }
    void stop() { std::cout << "Dvd player is stopped\n"; }
    void eject() { std::cout << "Dvd is ejected\n"; }
    void playMovie() { std::cout << "movie is being played\n"; }
};

class Amplifier {
   public:
    void on() { std::cout << "Amp is on\n"; }
    void off() { std::cout << "Amp is off\n"; }
    void setDvd(DvdPlayer dvd) { std::cout << "Used the dvd\n"; }
    void setSurroundSound() { std::cout << "have set the sound as surround\n"; }
    void setVolume(int volume) { std::cout << "the volume is " << volume << "\n"; }
};

class Projector {
   public:
    void on() { std::cout << "The projector is on\n"; }
    void off() { std::cout << "The projector is off\n"; }
    void wideScreenMode() { std::cout << "Switched to wide screen mode\n"; }
};

class Lights {
   public:
    void dim(int dimming) { std::cout << "Lights dimmed  to " << dimming << "%\n"; }
    void on() { std::cout << "The lights are on\n"; }
};

class Screen {
   public:
    void down() { std::cout << "screen is down\n"; }
    void up() { std::cout << "screen is up\n"; }
};

class PopcornPopper {
   public:
    void on() { std::cout << "Switched on Popper\n"; }
    void off() { std::cout << "Switched off Popper\n"; }
    void pop() { std::cout << "Popper stated poppig\n"; }
};

class HomeTheatreFacade {
    Amplifier amp;
    DvdPlayer dvd;
    Projector projector;
    Lights lights;
    Screen screen;
    PopcornPopper popper;

   public:
    explicit HomeTheatreFacade(Amplifier amp, DvdPlayer dvd, Projector projector, Lights lights, Screen screen, PopcornPopper popper)
        : amp(amp), dvd(dvd), projector(projector), lights(lights), screen(screen), popper(popper) {}

    void watchMovie() {
        popper.on();
        popper.pop();

        lights.dim(10);
        screen.down();
        projector.on();
        projector.wideScreenMode();
        amp.on();
        amp.setDvd(dvd);
        amp.setVolume(5);
        dvd.on();
        dvd.playMovie();
    }

    void endMovie() {
        popper.off();
        lights.on();
        screen.up();
        projector.off();
        amp.off();
        dvd.stop();
        dvd.eject();
        dvd.off();
    }
};

int main() {
	Amplifier amp;
    DvdPlayer dvd;
    Projector projector;
    Lights lights;
    Screen screen;
    PopcornPopper popper;
	HomeTheatreFacade f(amp, dvd, projector, lights, screen, popper);
	f.watchMovie();
	f.endMovie();
}

File: Facade_Pattern/fp.py


class DVDPlayer:
    def on(self) -> None: print("Dvd player is on")
    def off(self) -> None: print("Dvd player is off")
    def stop(self) -> None: print("Dvd player is stopped")
    def eject(self) -> None: print("Dvd is ejected")
    def playMovie(self) -> None: print("movie is being played")


class Amplifier:
    def on(self) -> None: print("Amp is on")
    def off(self) -> None: print("Amp is off")
    def setDVD(self) -> None: print("Used the dvd")
    def setSurroundSound(self) -> None: print("have set the sound as surround")
    def setVolume(self, volume: int) -> None: print(f"The volume is {volume}")

class Projector:
    def on(self) -> None: print("The projector is on")
    def off(self) -> None: print("The projector is off")
    def wideScreenMode(self) -> None: print("Switched to wide screen mode")

class Lights:
    def dim(self, dimming: int): print(f"Lights dimmed to {dimming}")
    def on(self) -> None: print("The lights are on")


class Screen:
    def down(self) -> None: print("screen is down")
    def up(self) -> None: print("screen is up")


class PopcornPopper:
    def on(self) -> None: print("Switched on Popper")
    def off(self) -> None: print("Switched off Popper")   
    def pop(self) -> None: print("Popper stated poppig")


class HomeTheatreFacade:
    def __init__(self, amp: Amplifier, dvd: DVDPlayer, projector: Projector, lights: Lights, screen: Screen, popper: PopcornPopper) -> None:
        self.amp: Amplifier = amp
        self.dvd: DVDPlayer = dvd
        self.projector: Projector = projector
        self.lights: Lights = lights
        self.screen: Screen = screen
        self.popper: PopcornPopper = popper

    def watchMovie(self) -> None:
        self.popper.on()
        self.popper.pop()
        self.lights.dim(10)
        self.screen.down()
        self.projector.on()
        self.projector.wideScreenMode()
        self.amp.on()
        self.amp.setDVD()
        self.amp.setVolume(5)
        self.dvd.on()
        self.dvd.playMovie()

    def endMovie(self):
        self.popper.off()
        self.lights.on()
        self.screen.up()
        self.projector.off()
        self.amp.off()
        self.dvd.stop()
        self.dvd.eject()
        self.dvd.off()

def main() -> None:
        f = HomeTheatreFacade(
            Amplifier(), DVDPlayer(), Projector(), Lights(), Screen(), PopcornPopper()
        )
        f.watchMovie()
        print("-----------------------")
        f.endMovie()

if __name__ == '__main__':
    main()
